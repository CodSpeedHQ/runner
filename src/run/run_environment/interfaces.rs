use serde::{Deserialize, Serialize};

#[derive(Deserialize, Serialize, Debug, Clone, PartialEq)]
#[serde(rename_all = "UPPERCASE")]
pub enum RepositoryProvider {
    GitLab,
    GitHub,
}

#[derive(Deserialize, Serialize, Debug, Clone, PartialEq)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum RunEnvironment {
    GithubActions,
    GitlabCi,
    Buildkite,
    Local,
}

#[derive(Deserialize, Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct RunEnvironmentMetadata {
    #[serde(rename = "ref")]
    pub ref_: String,
    pub head_ref: Option<String>,
    pub base_ref: Option<String>,
    pub owner: String,
    pub repository: String,
    pub event: RunEvent,
    pub sender: Option<Sender>,
    pub gh_data: Option<GhData>,
    pub gl_data: Option<GlData>,
    pub repository_root_path: String,
}

#[derive(Deserialize, Serialize, Debug, Clone, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum RunEvent {
    Push,
    PullRequest,
    WorkflowDispatch,
    Schedule,
    Local,
}

#[derive(Deserialize, Serialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct GhData {
    pub run_id: String,
    pub job: String,
}

#[derive(Deserialize, Serialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct GlData {
    pub run_id: String,
    pub job: String,
}

#[derive(Deserialize, Serialize, Debug, Clone, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct Sender {
    pub id: String,
    pub login: String,
}
