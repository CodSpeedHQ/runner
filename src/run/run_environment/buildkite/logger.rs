use crate::{
    logger::{get_group_event, get_json_event, GroupEvent},
    run::run_environment::logger::should_provider_logger_handle_record,
};
use log::*;
use simplelog::SharedLogger;
use std::{env, io::Write};

/// A logger that prints logs in the format expected by Buildkite
///
/// See https://buildkite.com/docs/pipelines/managing-log-output
pub struct BuildkiteLogger {
    log_level: LevelFilter,
}

impl BuildkiteLogger {
    pub fn new() -> Self {
        let log_level = env::var("CODSPEED_LOG")
            .ok()
            .and_then(|log_level| log_level.parse::<log::LevelFilter>().ok())
            .unwrap_or(log::LevelFilter::Info);
        Self { log_level }
    }
}

impl Log for BuildkiteLogger {
    fn enabled(&self, _metadata: &Metadata) -> bool {
        true
    }

    fn log(&self, record: &Record) {
        if !should_provider_logger_handle_record(record) {
            return;
        }

        let level = record.level();
        let message = record.args();

        if let Some(group_event) = get_group_event(record) {
            match group_event {
                GroupEvent::Start(name) => {
                    println!("--- {name}");
                }
                GroupEvent::StartOpened(name) => {
                    println!("+++ {name}");
                }
                GroupEvent::End => {}
            }
            return;
        }

        if get_json_event(record).is_some() {
            return;
        }

        if level > self.log_level {
            return;
        }
        // there is no support for log levels in Buildkite, so we print the level in the message
        match level {
            Level::Error => {
                println!("[ERROR] {message}");
            }
            Level::Warn => {
                println!("[WARN] {message}");
            }
            Level::Info => {
                println!("{message}");
            }
            Level::Debug => {
                println!("[DEBUG] {message}");
            }
            Level::Trace => {
                println!("[TRACE] {message}");
            }
        }
    }

    fn flush(&self) {
        std::io::stdout().flush().unwrap();
    }
}

impl SharedLogger for BuildkiteLogger {
    fn level(&self) -> LevelFilter {
        self.log_level
    }

    fn config(&self) -> Option<&simplelog::Config> {
        None
    }

    fn as_log(self: Box<Self>) -> Box<dyn Log> {
        Box::new(*self)
    }
}
