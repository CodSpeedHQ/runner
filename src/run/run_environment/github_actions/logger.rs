use crate::{
    logger::{GroupEvent, get_group_event, get_json_event},
    run::run_environment::logger::should_provider_logger_handle_record,
};
use log::*;
use simplelog::SharedLogger;
use std::{env, io::Write};

/// A logger that prints logs in the format expected by GitHub Actions, with grouping support.
///
/// See https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions
pub struct GithubActionLogger {
    log_level: LevelFilter,
}

impl GithubActionLogger {
    pub fn new() -> Self {
        // Only enable debug logging if it's enabled in GitHub Actions.
        // See: https://docs.github.com/en/actions/reference/workflows-and-actions/variables
        let log_level = if env::var("RUNNER_DEBUG").unwrap_or_default() == "1" {
            LevelFilter::Trace
        } else {
            env::var("CODSPEED_LOG")
                .ok()
                .and_then(|log_level| log_level.parse::<LevelFilter>().ok())
                .unwrap_or(LevelFilter::Info)
        };

        Self { log_level }
    }
}

impl Log for GithubActionLogger {
    fn enabled(&self, _metadata: &Metadata) -> bool {
        true
    }

    fn log(&self, record: &Record) {
        if !should_provider_logger_handle_record(record) {
            return;
        }

        let level = record.level();
        let message = record.args();

        if let Some(group_event) = get_group_event(record) {
            match group_event {
                GroupEvent::Start(name) | GroupEvent::StartOpened(name) => {
                    println!("::group::{name}");
                }
                GroupEvent::End => {
                    println!("::endgroup::");
                }
            }
            return;
        }

        if get_json_event(record).is_some() {
            return;
        }

        if level > self.log_level {
            return;
        }

        let prefix = match level {
            Level::Error => "::error::",
            Level::Warn => "::warning::",
            Level::Info => "",
            Level::Debug => "::debug::",
            Level::Trace => "::debug::[TRACE]",
        };
        let message_string = message.to_string();
        let lines = message_string.lines();
        // ensure that all the lines of the message have the prefix, otherwise GitHub Actions will not recognize the command for the whole string
        lines.for_each(|line| println!("{prefix}{line}"));
    }

    fn flush(&self) {
        std::io::stdout().flush().unwrap();
    }
}

impl SharedLogger for GithubActionLogger {
    fn level(&self) -> LevelFilter {
        self.log_level
    }

    fn config(&self) -> Option<&simplelog::Config> {
        None
    }

    fn as_log(self: Box<Self>) -> Box<dyn Log> {
        Box::new(*self)
    }
}
