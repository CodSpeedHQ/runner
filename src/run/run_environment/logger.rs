use crate::run::runner::EXECUTOR_TARGET;

pub(super) fn should_provider_logger_handle_record(record: &log::Record) -> bool {
    // Provider logger should handle all records except the ones from the executor target
    record.target() != EXECUTOR_TARGET
}
