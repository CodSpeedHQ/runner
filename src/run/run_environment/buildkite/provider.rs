use std::env;

use simplelog::SharedLogger;

use crate::prelude::*;
use crate::run::helpers::{parse_git_remote, GitRemote};
use crate::run::run_environment::interfaces::RunEnvironment;
use crate::run::{
    config::Config,
    helpers::{find_repository_root, get_env_variable},
    run_environment::{
        interfaces::{RepositoryProvider, RunEnvironmentMetadata, RunEvent},
        provider::{RunEnvironmentDetector, RunEnvironmentProvider},
    },
};

use super::logger::BuildkiteLogger;

#[derive(Debug)]
pub struct BuildkiteProvider {
    owner: String,
    repository: String,
    ref_: String,
    head_ref: Option<String>,
    base_ref: Option<String>,
    event: RunEvent,
    repository_root_path: String,
}

pub fn get_pr_number() -> Result<Option<u64>> {
    Ok(get_env_variable("BUILDKITE_PULL_REQUEST")?.parse().ok())
}

pub fn get_run_event() -> Result<RunEvent> {
    let is_pr = get_pr_number()?.is_some();

    if is_pr {
        Ok(RunEvent::PullRequest)
    } else {
        Ok(RunEvent::Push)
    }
}

pub fn get_ref() -> Result<String> {
    let pr_number = get_pr_number()?;

    if let Some(pr_number) = pr_number {
        Ok(format!("refs/pull/{}/merge", pr_number))
    } else {
        Ok(format!(
            "refs/heads/{}",
            get_env_variable("BUILDKITE_BRANCH")?
        ))
    }
}

impl TryFrom<&Config> for BuildkiteProvider {
    type Error = Error;
    fn try_from(config: &Config) -> Result<Self> {
        if config.token.is_none() {
            bail!("Token authentication is required for Buildkite");
        }

        let is_pr = get_pr_number()?.is_some();
        let repository_url = get_env_variable("BUILDKITE_REPO")?;
        let GitRemote {
            owner,
            repository,
            domain,
        } = parse_git_remote(&repository_url)?;

        if domain != "github.com" {
            bail!(
                "Only GitHub repositories are supported by CodSpeed BuildKite integration for now."
            );
        }

        let repository_root_path = match find_repository_root(&std::env::current_dir()?) {
            Some(mut path) => {
                // Add a trailing slash to the path
                path.push("");
                path.to_string_lossy().to_string()
            }
            None => format!(
                "/buildkite/builds/{}/{}/{}/",
                get_env_variable("BUILDKITE_AGENT_NAME")?,
                get_env_variable("BUILDKITE_ORGANIZATION_SLUG")?,
                get_env_variable("BUILDKITE_PIPELINE_SLUG")?,
            ),
        };

        Ok(Self {
            owner: owner.clone(),
            repository: repository.clone(),
            ref_: get_ref()?,
            base_ref: if is_pr {
                Some(get_env_variable("BUILDKITE_PULL_REQUEST_BASE_BRANCH")?)
            } else {
                None
            },
            head_ref: if is_pr {
                Some(get_env_variable("BUILDKITE_BRANCH")?)
            } else {
                None
            },
            event: get_run_event()?,
            repository_root_path,
        })
    }
}

impl RunEnvironmentDetector for BuildkiteProvider {
    fn detect() -> bool {
        env::var("BUILDKITE") == Ok("true".into())
    }
}

impl RunEnvironmentProvider for BuildkiteProvider {
    fn get_repository_provider(&self) -> RepositoryProvider {
        RepositoryProvider::GitHub
    }

    fn get_logger(&self) -> Box<dyn SharedLogger> {
        Box::new(BuildkiteLogger::new())
    }

    fn get_run_environment_name(&self) -> &'static str {
        "Buildkite"
    }

    fn get_run_environment(&self) -> RunEnvironment {
        RunEnvironment::Buildkite
    }

    fn get_run_environment_metadata(&self) -> Result<RunEnvironmentMetadata> {
        Ok(RunEnvironmentMetadata {
            base_ref: self.base_ref.clone(),
            head_ref: self.head_ref.clone(),
            event: self.event.clone(),
            owner: self.owner.clone(),
            repository: self.repository.clone(),
            ref_: self.ref_.clone(),
            repository_root_path: self.repository_root_path.clone(),
            gh_data: None,
            gl_data: None,
            sender: None,
        })
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_json_snapshot;
    use temp_env::{with_var, with_vars};

    use crate::VERSION;

    use super::*;

    #[test]
    fn test_detect() {
        with_var("BUILDKITE", Some("true"), || {
            assert!(BuildkiteProvider::detect());
        });
    }

    #[test]
    fn test_try_from_push_main() {
        with_vars(
            [
                ("BUILDKITE_AGENT_NAME", Some("7b10eca7600b-1")),
                ("BUILDKITE_BRANCH", Some("main")),
                ("BUILDKITE_BUILD_NUMBER", Some("1")),
                ("BUILDKITE_COMMIT", Some("abc123")),
                ("BUILDKITE_ORGANIZATION_SLUG", Some("my-org")),
                ("BUILDKITE_PIPELINE_SLUG", Some("buildkite-test")),
                ("BUILDKITE_PULL_REQUEST_BASE_BRANCH", Some("")),
                ("BUILDKITE_PULL_REQUEST", Some("")),
                (
                    "BUILDKITE_REPO",
                    Some("https://github.com/my-org/adrien-python-test.git"),
                ),
                ("BUILDKITE", Some("true")),
            ],
            || {
                let config = Config {
                    token: Some("token".into()),
                    ..Config::test()
                };
                let provider = BuildkiteProvider::try_from(&config).unwrap();

                assert_eq!(provider.owner, "my-org");
                assert_eq!(provider.repository, "adrien-python-test");
                assert_eq!(provider.ref_, "refs/heads/main");
                assert_eq!(provider.base_ref, None);
                assert_eq!(provider.head_ref, None);
                assert_eq!(provider.event, RunEvent::Push);
                assert_eq!(
                    provider.repository_root_path,
                    "/buildkite/builds/7b10eca7600b-1/my-org/buildkite-test/"
                );
            },
        );
    }

    #[test]
    fn test_try_from_pull_request() {
        with_vars(
            [
                ("BUILDKITE_AGENT_NAME", Some("7b10eca7600b-1")),
                ("BUILDKITE_BRANCH", Some("feat/codspeed-runner")),
                ("BUILDKITE_BUILD_NUMBER", Some("1")),
                ("BUILDKITE_COMMIT", Some("abc123")),
                ("BUILDKITE_ORGANIZATION_SLUG", Some("my-org")),
                ("BUILDKITE_PIPELINE_SLUG", Some("buildkite-test")),
                ("BUILDKITE_PULL_REQUEST_BASE_BRANCH", Some("main")),
                ("BUILDKITE_PULL_REQUEST", Some("22")),
                (
                    "BUILDKITE_REPO",
                    Some("git@github.com:my-org/adrien-python-test.git"),
                ),
                ("BUILDKITE", Some("true")),
            ],
            || {
                let config = Config {
                    token: Some("token".into()),
                    ..Config::test()
                };
                let provider = BuildkiteProvider::try_from(&config).unwrap();

                assert_eq!(provider.owner, "my-org");
                assert_eq!(provider.repository, "adrien-python-test");
                assert_eq!(provider.ref_, "refs/pull/22/merge");
                assert_eq!(provider.base_ref, Some("main".into()));
                assert_eq!(provider.head_ref, Some("feat/codspeed-runner".into()));
                assert_eq!(provider.event, RunEvent::PullRequest);
                assert_eq!(
                    provider.repository_root_path,
                    "/buildkite/builds/7b10eca7600b-1/my-org/buildkite-test/"
                );
            },
        );
    }

    #[test]
    fn test_pull_request_run_environment_metadata() {
        with_vars(
            [
                ("BUILDKITE_AGENT_NAME", Some("7b10eca7600b-1")),
                ("BUILDKITE_BRANCH", Some("feat/codspeed-runner")),
                ("BUILDKITE_BUILD_NUMBER", Some("1")),
                ("BUILDKITE_COMMIT", Some("abc123")),
                ("BUILDKITE_ORGANIZATION_SLUG", Some("my-org")),
                ("BUILDKITE_PIPELINE_SLUG", Some("buildkite-test")),
                ("BUILDKITE_PULL_REQUEST_BASE_BRANCH", Some("main")),
                ("BUILDKITE_PULL_REQUEST", Some("22")),
                (
                    "BUILDKITE_REPO",
                    Some("git@github.com:my-org/adrien-python-test.git"),
                ),
                ("BUILDKITE", Some("true")),
            ],
            || {
                let config = Config {
                    token: Some("token".into()),
                    ..Config::test()
                };
                let provider = BuildkiteProvider::try_from(&config).unwrap();
                let run_environment_metadata = provider.get_run_environment_metadata().unwrap();

                assert_json_snapshot!(run_environment_metadata, {
                    ".runner.version" => insta::dynamic_redaction(|value,_path| {
                        assert_eq!(value.as_str().unwrap(), VERSION.to_string());
                        "[version]"
                    }),
                })
            },
        );
    }
}
