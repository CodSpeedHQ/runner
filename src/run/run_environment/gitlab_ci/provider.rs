use simplelog::SharedLogger;
use std::env;

use crate::prelude::*;
use crate::run::config::Config;
use crate::run::helpers::get_env_variable;
use crate::run::run_environment::interfaces::{
    GlData, RepositoryProvider, RunEnvironment, RunEnvironmentMetadata, RunEvent, Sender,
};
use crate::run::run_environment::provider::RunEnvironmentDetector;
use crate::run::run_environment::RunEnvironmentProvider;

use super::logger::GitLabCILogger;

#[derive(Debug)]
pub struct GitLabCIProvider {
    owner: String,
    repository: String,
    ref_: String,
    head_ref: Option<String>,
    base_ref: Option<String>,
    gl_data: GlData,
    sender: Sender,
    event: RunEvent,
    repository_root_path: String,
}

impl TryFrom<&Config> for GitLabCIProvider {
    type Error = Error;
    fn try_from(_config: &Config) -> Result<Self> {
        let owner = get_env_variable("CI_PROJECT_NAMESPACE")?;
        let repository = get_env_variable("CI_PROJECT_NAME")?;

        let ci_pipeline_source = get_env_variable("CI_PIPELINE_SOURCE")?;

        let branch_name = get_env_variable("CI_COMMIT_REF_NAME")?;

        // compute the branch or tag ref which mimics GitHub behavior
        // CI_COMMIT_TAG is only present in pipelines for tags.
        // See https://docs.gitlab.com/ee/ci/variables/predefined_variables.html
        let branch_or_tag_ref = get_env_variable("CI_COMMIT_TAG")
            .map(|tag_name| format!("refs/tags/{tag_name}"))
            .unwrap_or(format!("refs/heads/{branch_name}"));

        // https://docs.gitlab.com/ee/ci/jobs/job_rules.html#ci_pipeline_source-predefined-variable
        let (event, ref_, base_ref, head_ref) = match ci_pipeline_source.as_str() {
            // For pipelines created when a merge request is created or updated. Required to enable merge request pipelines, merged results pipelines, and merge trains.
            // https://docs.gitlab.com/ee/ci/variables/predefined_variables.html#predefined-variables-for-merge-request-pipelines
            "merge_request_event" => {
                let merge_request_id = get_env_variable("CI_MERGE_REQUEST_IID")?;
                let target_branch_name = get_env_variable("CI_MERGE_REQUEST_TARGET_BRANCH_NAME")?;
                let source_branch_name = get_env_variable("CI_MERGE_REQUEST_SOURCE_BRANCH_NAME")?;

                // check if the merge request is from a fork
                let ci_project_path = get_env_variable("CI_PROJECT_PATH")?;
                let ci_merge_request_source_project_path =
                    get_env_variable("CI_MERGE_REQUEST_SOURCE_PROJECT_PATH")?;

                if ci_project_path != ci_merge_request_source_project_path {
                    let fork_owner = ci_merge_request_source_project_path
                        .split('/')
                        .next()
                        .expect("Malformed Source Project Path");

                    (
                        RunEvent::PullRequest,
                        format!("refs/pull/{merge_request_id}/merge"),
                        Some(target_branch_name),
                        Some(format!("{fork_owner}:{source_branch_name}")),
                    )
                } else {
                    (
                        RunEvent::PullRequest,
                        format!("refs/pull/{merge_request_id}/merge"),
                        Some(target_branch_name),
                        Some(source_branch_name),
                    )
                }
            }

            // For pipelines triggered by a Git push event, including for branches and tags.
            "push" => (RunEvent::Push, branch_or_tag_ref, Some(branch_name), None),

            // For scheduled pipelines.
            "schedule" => (
                RunEvent::Schedule,
                branch_or_tag_ref,
                Some(branch_name),
                None,
            ),

            // For pipelines created with the api, using a trigger token or via the GitLab UI.
            "trigger" | "web" | "api" => (
                RunEvent::WorkflowDispatch,
                branch_or_tag_ref,
                Some(branch_name),
                None,
            ),

            _ => bail!("Event {} is not supported by CodSpeed", ci_pipeline_source),
        };

        let run_id = get_env_variable("CI_JOB_ID")?;
        let job = get_env_variable("CI_JOB_NAME")?;

        let gitlab_user_id = get_env_variable("GITLAB_USER_ID")?;
        let gitlab_user_login = get_env_variable("GITLAB_USER_LOGIN")?;

        let gl_data = GlData { run_id, job };
        let sender = Sender {
            id: gitlab_user_id,
            login: gitlab_user_login,
        };

        let repository_root_path = get_env_variable("CI_PROJECT_DIR")?;

        Ok(Self {
            owner,
            repository,
            ref_,
            head_ref,
            base_ref,
            gl_data,
            sender,
            event,
            repository_root_path,
        })
    }
}

impl RunEnvironmentDetector for GitLabCIProvider {
    fn detect() -> bool {
        // check if the GITLAB_CI environment variable is set and the value is truthy
        env::var("GITLAB_CI") == Ok("true".into())
    }
}

impl RunEnvironmentProvider for GitLabCIProvider {
    fn get_logger(&self) -> Box<dyn SharedLogger> {
        Box::new(GitLabCILogger::new())
    }

    fn get_repository_provider(&self) -> RepositoryProvider {
        RepositoryProvider::GitLab
    }

    fn get_run_environment_name(&self) -> &'static str {
        "GitLab CI"
    }

    fn get_run_environment(&self) -> RunEnvironment {
        RunEnvironment::GitlabCi
    }

    fn get_run_environment_metadata(&self) -> Result<RunEnvironmentMetadata> {
        Ok(RunEnvironmentMetadata {
            base_ref: self.base_ref.clone(),
            head_ref: self.head_ref.clone(),
            event: self.event.clone(),
            gh_data: None,
            gl_data: Some(self.gl_data.clone()),
            sender: Some(self.sender.clone()),
            owner: self.owner.clone(),
            repository: self.repository.clone(),
            ref_: self.ref_.clone(),
            repository_root_path: self.repository_root_path.clone(),
        })
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_json_snapshot;
    use temp_env::{with_var, with_vars};

    use crate::VERSION;

    use super::*;

    #[test]
    fn test_detect() {
        with_var("GITLAB_CI", Some("true"), || {
            assert!(GitLabCIProvider::detect());
        });
    }

    #[test]
    fn test_push_main_run_environment_metadata() {
        with_vars(
            [
                ("GITLAB_CI", Some("true")),
                ("CI_PROJECT_DIR", Some("/builds/owner/repository")),
                ("GITLAB_USER_ID", Some("1234567890")),
                ("GITLAB_USER_LOGIN", Some("actor")),
                ("CI_PROJECT_NAME", Some("repository")),
                ("CI_PROJECT_NAMESPACE", Some("owner")),
                ("CI_JOB_NAME", Some("job")),
                ("CI_JOB_ID", Some("1234567890")),
                ("CI_PIPELINE_SOURCE", Some("push")),
                ("CI_COMMIT_REF_NAME", Some("main")),
            ],
            || {
                let config = Config {
                    token: Some("token".into()),
                    ..Config::test()
                };
                let gitlab_ci_provider = GitLabCIProvider::try_from(&config).unwrap();
                let run_environment_metadata =
                    gitlab_ci_provider.get_run_environment_metadata().unwrap();

                assert_json_snapshot!(run_environment_metadata, {
                    ".runner.version" => insta::dynamic_redaction(|value,_path| {
                        assert_eq!(value.as_str().unwrap(), VERSION.to_string());
                        "[version]"
                    }),
                });
            },
        )
    }

    #[test]
    fn test_merge_request_run_environment_metadata() {
        with_vars(
            [
                ("GITLAB_CI", Some("true")),
                ("CI_PROJECT_DIR", Some("/builds/owner/repository")),
                ("GITLAB_USER_ID", Some("19605940")),
                ("GITLAB_USER_LOGIN", Some("actor")),
                ("CI_PROJECT_NAME", Some("repository")),
                ("CI_PROJECT_NAMESPACE", Some("owner")),
                ("CI_JOB_NAME", Some("build-job")),
                ("CI_JOB_ID", Some("6957110437")),
                ("CI_PIPELINE_SOURCE", Some("merge_request_event")),
                ("CI_COMMIT_REF_NAME", Some("main")),
                ("CI_MERGE_REQUEST_IID", Some("5")),
                ("CI_MERGE_REQUEST_TARGET_BRANCH_NAME", Some("main")),
                (
                    "CI_MERGE_REQUEST_SOURCE_BRANCH_NAME",
                    Some("feat/awesome-feature"),
                ),
                ("CI_PROJECT_PATH", Some("owner/repository")),
                (
                    "CI_MERGE_REQUEST_SOURCE_PROJECT_PATH",
                    Some("owner/repository"),
                ),
            ],
            || {
                let config = Config {
                    token: Some("token".into()),
                    ..Config::test()
                };
                let gitlab_ci_provider = GitLabCIProvider::try_from(&config).unwrap();
                let run_environment_metadata =
                    gitlab_ci_provider.get_run_environment_metadata().unwrap();

                assert_json_snapshot!(run_environment_metadata, {
                    ".runner.version" => insta::dynamic_redaction(|value,_path| {
                        assert_eq!(value.as_str().unwrap(), VERSION.to_string());
                        "[version]"
                    }),
                });
            },
        );
    }

    #[test]
    fn test_fork_merge_request_run_environment_metadata() {
        with_vars(
            [
                ("GITLAB_CI", Some("true")),
                ("CI_PROJECT_DIR", Some("/builds/owner/repository")),
                ("GITLAB_USER_ID", Some("19605940")),
                ("GITLAB_USER_LOGIN", Some("actor")),
                ("CI_PROJECT_NAME", Some("repository")),
                ("CI_PROJECT_NAMESPACE", Some("owner")),
                ("CI_JOB_NAME", Some("build-job")),
                ("CI_JOB_ID", Some("6957110437")),
                ("CI_PIPELINE_SOURCE", Some("merge_request_event")),
                ("CI_COMMIT_REF_NAME", Some("main")),
                ("CI_MERGE_REQUEST_IID", Some("5")),
                ("CI_MERGE_REQUEST_TARGET_BRANCH_NAME", Some("main")),
                (
                    "CI_MERGE_REQUEST_SOURCE_BRANCH_NAME",
                    Some("feat/awesome-feature"),
                ),
                ("CI_PROJECT_PATH", Some("owner/repository")),
                (
                    "CI_MERGE_REQUEST_SOURCE_PROJECT_PATH",
                    Some("fork-owner/fork-repository"),
                ),
            ],
            || {
                let config = Config {
                    token: Some("token".into()),
                    ..Config::test()
                };
                let gitlab_ci_provider = GitLabCIProvider::try_from(&config).unwrap();
                let run_environment_metadata =
                    gitlab_ci_provider.get_run_environment_metadata().unwrap();

                assert_json_snapshot!(run_environment_metadata, {
                    ".runner.version" => insta::dynamic_redaction(|value,_path| {
                        assert_eq!(value.as_str().unwrap(), VERSION.to_string());
                        "[version]"
                    }),
                });
            },
        );
    }
}
