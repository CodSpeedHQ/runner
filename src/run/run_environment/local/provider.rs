use git2::Repository;
use simplelog::SharedLogger;

use crate::local_logger::get_local_logger;
use crate::prelude::*;
use crate::run::helpers::{parse_git_remote, GitRemote};
use crate::run::run_environment::interfaces::RunEnvironment;
use crate::run::{
    config::Config,
    helpers::find_repository_root,
    run_environment::{
        interfaces::{RepositoryProvider, RunEnvironmentMetadata, RunEvent},
        provider::{RunEnvironmentDetector, RunEnvironmentProvider},
    },
};

#[derive(Debug)]
pub struct LocalProvider {
    repository_provider: RepositoryProvider,
    pub ref_: String,
    pub owner: String,
    pub repository: String,
    pub head_ref: Option<String>,
    pub base_ref: Option<String>,
    pub event: RunEvent,
    pub repository_root_path: String,
}

impl LocalProvider {}

fn extract_provider_owner_and_repository_from_remote_url(
    remote_url: &str,
) -> Result<(RepositoryProvider, String, String)> {
    let GitRemote {
        domain,
        owner,
        repository,
    } = parse_git_remote(remote_url)?;
    let repository_provider = match domain.as_str() {
        "github.com" => RepositoryProvider::GitHub,
        "gitlab.com" => RepositoryProvider::GitLab,
        domain => bail!(
            "Repository provider {} is not supported by CodSpeed",
            domain
        ),
    };

    Ok((
        repository_provider,
        owner.to_string(),
        repository.to_string(),
    ))
}

impl TryFrom<&Config> for LocalProvider {
    type Error = Error;
    fn try_from(_config: &Config) -> Result<Self> {
        let repository_root_path = match find_repository_root(&std::env::current_dir()?) {
            Some(mut path) => {
                // Add a trailing slash to the path
                path.push("");
                path.to_string_lossy().to_string()
            },
            None => bail!("Could not find repository root, please make sure you are running the command from inside a git repository"),
        };

        let git_repository = Repository::open(repository_root_path.clone()).context(format!(
            "Failed to open repository at path: {}",
            repository_root_path
        ))?;

        let remote = git_repository.find_remote("origin")?;
        let (repository_provider, owner, repository) =
            extract_provider_owner_and_repository_from_remote_url(remote.url().unwrap())?;

        let head = git_repository.head().context("Failed to get HEAD")?;
        let ref_ = head
            .peel_to_commit()
            .context("Failed to get HEAD commit")?
            .id()
            .to_string();
        let head_ref = if head.is_branch() {
            let branch = head.shorthand().context("Failed to get HEAD branch name")?;
            Some(branch.to_string())
        } else {
            None
        };

        Ok(Self {
            repository_provider,
            ref_,
            head_ref,
            base_ref: None,
            owner,
            repository,
            event: RunEvent::Local,
            repository_root_path,
        })
    }
}

impl RunEnvironmentDetector for LocalProvider {
    fn detect() -> bool {
        true
    }
}

impl RunEnvironmentProvider for LocalProvider {
    fn get_repository_provider(&self) -> RepositoryProvider {
        self.repository_provider.clone()
    }

    fn get_logger(&self) -> Box<dyn SharedLogger> {
        get_local_logger()
    }

    fn get_run_environment(&self) -> RunEnvironment {
        RunEnvironment::Local
    }

    fn get_run_environment_metadata(&self) -> Result<RunEnvironmentMetadata> {
        Ok(RunEnvironmentMetadata {
            base_ref: self.base_ref.clone(),
            head_ref: self.head_ref.clone(),
            event: self.event.clone(),
            gh_data: None,
            gl_data: None,
            sender: None,
            owner: self.owner.clone(),
            repository: self.repository.clone(),
            ref_: self.ref_.clone(),
            repository_root_path: self.repository_root_path.clone(),
        })
    }
}

#[cfg(test)]
mod tests {
    // use crate::VERSION;
    // use insta::assert_json_snapshot;

    use super::*;

    #[test]
    fn test_extract_provider_owner_and_repository_from_remote_url() {
        let remote_urls = [
            (
                "git@github.com:CodSpeedHQ/runner.git",
                RepositoryProvider::GitHub,
                "CodSpeedHQ",
                "runner",
            ),
            (
                "https://github.com/CodSpeedHQ/runner.git",
                RepositoryProvider::GitHub,
                "CodSpeedHQ",
                "runner",
            ),
            (
                "git@gitlab.com:codspeed/runner.git",
                RepositoryProvider::GitLab,
                "codspeed",
                "runner",
            ),
            (
                "https://gitlab.com/codspeed/runner.git",
                RepositoryProvider::GitLab,
                "codspeed",
                "runner",
            ),
        ];
        for (remote_url, expected_provider, expected_owner, expected_repository) in
            remote_urls.into_iter()
        {
            let (repository_provider, owner, repository) =
                extract_provider_owner_and_repository_from_remote_url(remote_url).unwrap();
            assert_eq!(repository_provider, expected_provider);
            assert_eq!(owner, expected_owner);
            assert_eq!(repository, expected_repository);
        }
    }

    // TODO: uncomment later when we have a way to mock git repository
    // #[test]
    // fn test_provider_metadata() {
    //     let config = Config {
    //         token: Some("token".into()),
    //         ..Config::test()
    //     };
    //     let local_provider = LocalProvider::try_from(&config).unwrap();
    //     let provider_metadata = local_provider.get_provider_metadata().unwrap();

    //     assert_json_snapshot!(provider_metadata, {
    //         ".runner.version" => insta::dynamic_redaction(|value,_path| {
    //             assert_eq!(value.as_str().unwrap(), VERSION.to_string());
    //             "[version]"
    //         }),
    //     });
    // }
}
