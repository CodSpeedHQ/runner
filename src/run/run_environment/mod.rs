pub mod interfaces;
pub mod logger;
mod provider;

use buildkite::BuildkiteProvider;
use github_actions::GitHubActionsProvider;
use gitlab_ci::GitLabCIProvider;
use local::LocalProvider;
use provider::RunEnvironmentDetector;

use crate::prelude::*;
use crate::run::config::Config;

pub use self::interfaces::*;
pub use self::provider::RunEnvironmentProvider;

// RunEnvironment Provider implementations
mod buildkite;
mod github_actions;
mod gitlab_ci;
mod local;

pub fn get_provider(config: &Config) -> Result<Box<dyn RunEnvironmentProvider>> {
    if BuildkiteProvider::detect() {
        let provider = BuildkiteProvider::try_from(config)?;
        return Ok(Box::new(provider));
    }

    if GitHubActionsProvider::detect() {
        let provider = GitHubActionsProvider::try_from(config)?;
        return Ok(Box::new(provider));
    }

    if GitLabCIProvider::detect() {
        let provider = GitLabCIProvider::try_from(config)?;
        return Ok(Box::new(provider));
    }

    if LocalProvider::detect() {
        let provider = LocalProvider::try_from(config)?;
        return Ok(Box::new(provider));
    }

    // By design, this should not happen as the `LocalProvider` is a fallback
    bail!("No RunEnvironment provider detected")
}
