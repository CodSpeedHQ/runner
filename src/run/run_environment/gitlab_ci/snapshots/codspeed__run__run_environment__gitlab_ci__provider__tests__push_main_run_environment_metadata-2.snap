---
source: src/run/run_environment/gitlab_ci/provider.rs
expression: run_part
---
{
  "runId": "1234567890",
  "runPartId": "job",
  "jobName": "job",
  "metadata": {}
}
